### ISSUE TITLE
WebAssembly bindings API to provide tree-sitter.wasm location

### ISSUE THREAD

**Issue by drom** [tree-sitter/tree-sitter#559]
While using WebAssembly bindings

as described here: https://github.com/tree-sitter/tree-sitter/tree/master/lib/binding_web

```
const Parser = require('web-tree-sitter');
Parser.init().then(() => { /* the library is ready */ });
```

I noticed that `Parser.init()` method trying to fetch `tree-sitter.wasm` file from the path relative to the current page location. Can we provide `fetch` location to the `inits()` method?

**Comment by maxbrunsfeld** [comment]
Just to clarify, I don't think it's based on the current page's URL. It's based on the URL that the `tree-sitter.js` file was loaded from. This is actually controlled by Emscripten, not something that's explicitly controlled in the library, but there might be a way to customize it; I'm not sure.

If there is a way, I'd be open to adding an optional argument to `Parser.init()`, for specifying the `.wasm` URL. Are you interested in working on that?

**Comment by drom** [comment]
Here is my use case: https://observablehq.com/@drom/tree-sitter

**Comment by drom** [comment]
Ok, I found `script/build-wasm` and can build `tree-sitter{js,wasm}`

**Comment by drom** [comment]
I get it to work here https://observablehq.com/@drom/tree-sitter by patching `prefix.js`: 

```
Module.locateFile = function (url) {
  return 'https://raw.githubusercontent.com/chipsalliance/tree-sitter-firrtl/master/docs/' + url;
};
```
Name `tree-sitter.wasm` got hardcoded in `tree-sitter.js`.
Full path resolution or `tree-sitter.wasm` happens in the constructor of a class.
Fetch of `tree-sitter.wasm` happens before the `init()`

**Comment by Menci** [comment]
Without a function providing the WASM location, the module won't work in Webpack.

**Comment by Menci** [comment]
Another workaround (without patching web-tree-sitter) is patching the `fetch` function:

```js
function patchFetch() {
  const realFetch = window.fetch;
  window.fetch = function () {
    if (arguments[0].endsWith("/tree-sitter.wasm")) arguments[0] = TreeSitterWasmUrl;
    return realFetch.apply(window, arguments);
  };
}
```

**Comment by benlongo** [comment]
One possibility is to utilize `-s SINGLE_FILE=1` for emscripten. This embeds everything (base64 encoded) into the js output. However, this does increase artifact size by ~33%.

**Comment by Menci** [comment]
No, don't encode the WASM file to the JS file. BASE64 increased the bundle size for nothing.

**Comment by Menci** [comment]
@drom Yes, the `Parser.init` is called after the emscripten's glue code start trying to load the WASM file. So we need some changes to the structure.

#569 does some restructure to make use of emscripten's modularize option, it will be not hard to add `Parser.init(wasmPath)` support after that. See https://github.com/tree-sitter/tree-sitter/pull/569#issuecomment-596058030.

**Comment by milahu** [comment]
> ```js
>     if (arguments[0].endsWith("/tree-sitter.wasm")) arguments[0] = TreeSitterWasmUrl;
> ```

fix: `arguments[0] == "tree-sitter.wasm"`

**Comment by samestep** [comment]
Seems like this is still an issue, given that #569 was never merged; @Menci is there no easy way to modify the current build to use `-s MODULARIZE=1`?

**Comment by Enteleform** [comment]
+1

I've run into two contexts where the implicit `tree-sitter.wasm` location causes errors:

- monorepos with symlinked modules
  - e.g. a [Yarn workspace](https://classic.yarnpkg.com/lang/en/docs/workspaces/)
- environments that require a single-file bundle
  - e.g. an [Obsidian plugin](https://github.com/obsidianmd/obsidian-sample-plugin)

**Comment by amaanq** [comment]
https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/README.md#loading-the-wasm-file
